core/sail/lmdb/src/main/java/org/eclipse/rdf4j/sail/lmdb/MemoryOverflowModel.java

```diff
@@ -271,26 +265,7 @@ private synchronized void overflowToDisk() {
 			dataDir = Files.createTempDirectory("model").toFile();
 			logger.debug("memory overflow using temp directory {}", dataDir);
 			store = createSailStore(dataDir);
-			disk = new SailSourceModel(store) {
-
-				@Override
-				protected void finalize() throws Throwable {
-					logger.debug("finalizing {}", dataDir);
-					if (disk == this) {
-						try {
-							store.close();
-						} catch (SailException e) {
-							logger.error(e.toString(), e);
-						} finally {
-							FileUtil.deleteDir(dataDir);
-							dataDir = null;
-							store = null;
-							disk = null;
-						}
-					}
-					super.finalize();
-				}
-			};
```

Comment:
I would leave this code in, or copy the code from the MemoryOverflowIntoNativeStore that uses the Java 9 Cleaner. I had previously found a few issues with the code that handles closing of a ChangeSet, it's not exactly straight forward where and when a ChangeSet ends up being closed and there might still be bugs with how it's handled.

Reply:
Then I would propose to implement finalize on `ChangeSet`.

Reply:
I would rely on the fact that every ChangeSet is closed. Everything else means additional overhead for tracking the objects when we do not want to use `finalize`.

Reply:
The overhead of the Java 9 Cleaner is very low, especially compared to finalize. I have been using [ConcurrentCleaner](https://github.com/eclipse/rdf4j/blob/319c76d04abb8a0c0c39efdae540a4ed2ce19042/core/sail/api/src/main/java/org/eclipse/rdf4j/common/concurrent/locks/diagnostics/ConcurrentCleaner.java#L23) in a few places already. It distributes the load of registering between 128 Cleaners based on the thread id.

Reply:
I would be weary of putting too much trust to the changeset getting closed. There may still be bugs with that.